### ISSUE TITLE
Addressing the uncertainty in Reporting API

### ISSUE THREAD

**Issue by jugglinmike** [web-platform-tests/wpt#24327]
[The Reporting API does not make any guarantee about if or when reports are delivered via HTTP](https://w3c.github.io/reporting/#report-delivery), making it difficult to write authoritative tests against the feature.

Despite this, tests have already been written against this interface. For example: [CSP](https://github.com/web-platform-tests/wpt/blob/8e2d134568150ab920a8ddf303a2f0ef7a7cba6d/content-security-policy/reporting/report-and-enforce.html), [COEP](https://github.com/web-platform-tests/wpt/blob/8e2d134568150ab920a8ddf303a2f0ef7a7cba6d/html/cross-origin-embedder-policy/reporting-to-endpoint.https.html), and [COOP](https://github.com/web-platform-tests/wpt/blob/8e2d134568150ab920a8ddf303a2f0ef7a7cba6d/html/cross-origin-opener-policy/reporting-coop-navigated-popup.https.html). This suggests that each UA's default configuration is satisfactory for implementers, but it still leaves test writers in the dark about what constitutes a reasonable test.

It would be better if WPT acknowledged this uncertainty and documented how to mitigate the risks arising from it. Two solutions come to mind:

- extend [the Assumptions document](http://web-platform-tests.org/writing-tests/assumptions.html) with a specific duration after which browsers are expected to have delivered reports. Learn if this behavior can be configured in any of the browsers from which we currently collect results, and if so, update the infrastructure accordingly
- mark tests concerning HTTP report delivery as "optional" via [file name flag](http://web-platform-tests.org/writing-tests/file-names.html)

@dcreager @clelland @mikewest @camillelamy do any of you folks have thoughts about how we should address this?

**Comment by dcreager** [comment]
For Chromium, the Reporting API's "delivery interval" is configurable, along with a [few other settings](https://source.chromium.org/chromium/chromium/src/+/master:net/reporting/reporting_policy.proto).  When running WPT tests, that interval is [set to 100ms](https://source.chromium.org/chromium/chromium/src/+/master:content/shell/browser/shell_content_browser_client.h;l=219-221).  Then (at least for NEL), each test case waits a small amount of time ([typically 1 second](https://github.com/web-platform-tests/wpt/blob/master/network-error-logging/support/nel.sub.js#L201-L205)) and only declares a test as failing if a report doesn't arrive in that interval.  That sounds like something that we could codify in the Assumptions doc like you suggest.

**Comment by jugglinmike** [comment]
Thanks @dcreager, that sounds promising

@Hexcles, `--run-web-tests` doesn't show up in WPT. Is that only available in content_shell? Do you know if we'll be able to enable similar behavior for wpt.fyi? (Also, hi Robert!)

**Comment by Hexcles** [comment]
That's right, `--run-web-tests` isn't available outside of content_shell. Is that a requirement for these tests? Could we expose the necessary settings via Chrome CLI flags and/or ChromeDriver prefs?

**Comment by jugglinmike** [comment]
> Is that a requirement for these tests?

Reporting is a cross-cutting feature, so it's tough to answer this generally.

- [Chrome is failing the NEL tests on wpt.fyi](https://wpt.fyi/results/network-error-logging?run_id=584960003&run_id=608080002&run_id=596800001&run_id=596780004), so it seems necessary there.
- [Chrome is doing just fine on the CSP reporting tests](https://wpt.fyi/results/content-security-policy/reporting?run_id=584960003&run_id=608080002&run_id=596800001&run_id=596780004), suggesting that the NEL tests may have some other problem.
- [Chrome is also failing the COOP reporting tests on wpt.fyi](https://wpt.fyi/results/html/cross-origin-opener-policy?run_id=584960003&run_id=608080002&run_id=596800001&run_id=596780004), but that's a weaker signal because the spec is in flux.

> Could we expose the necessary settings via Chrome CLI flags and/or ChromeDriver prefs?

That's what I had in mind. Does it seem feasible?

An alternative would be some sort of WebDriver setting, but that would take even more effort. Since this isn't really useful for web developers, the overhead doesn't seem warranted.
